init.lua

```diff
@@ -137,23 +141,64 @@ function can_build_rail(start_pos, end_pos)
 	return false
 end
 
+local function can_dig(pos, player)
+	return (minetest.registered_nodes[advtrains.ndb.get_node(pos).name].can_dig or function() return true end)(pos, player)
+end
+
+--[[
+Checks if it is "appropriate" to change the node at <pos> to <newnode>
+Returns: new node (may be adjusted for e.g. turnouts or crossings), error message for the player
+--]]
+local function can_overwrite_track_at(player, pos, newnode)
+	local oldnode = advtrains.ndb.get_node(pos)
+	local oldname, oldparam2, newname, newparam2 = oldnode.name, oldnode.param2, newnode.name, newnode.param2
+	if not can_dig(pos, player) then
+		return false, S("Cannot modify the node at @1", minetest.pos_to_string(pos))
+	end
+	if oldname == newname and oldparam2 == newparam2 then
+		return false
+	end
+	if not advtrains.is_track_and_drives_on(oldname) then
+		return newnode
+	end
+	if not advtrains.is_track_and_drives_on(newname) then
+		return false, S("@1 is not a track", quote_string(newname))
+	end
+	local oldconns, _, oldtype = advtrains.get_track_connections(oldname, oldparam2)
+	local newconns, _, newtype = advtrains.get_track_connections(newname, newparam2)
+	if oldtype ~= newtype then
+		return false, S("The track at @1 is of a different type than @2", minetest.pos_to_string(pos), quote_string(newname))
+	elseif #oldconns > #newconns then
+		return false, S("@1 would not fit the track at @2", quote_string(newname), minetest.pos_to_string(pos))
+	end
+	if oldconns[3] and newconns[3] and type(oldconns[5]) ~= type(newconns[5]) then
+		return false, S("@1 would not fit the track at @2", quote_string(newname), minetest.pos_to_string(pos))
+	end
+	return newnode
+end
+
+local function try_build(player, pos, newnode)
+	local place_current, errmsg = can_overwrite_track_at(player, pos, newnode)
+	if place_current then
+		advtrains.ndb.swap_node(pos, place_current)
+	end
+	if errmsg then
+		minetest.chat_send_player(player:get_player_name(), errmsg)
+	end
+	return place_current
+end
+
 -- build a rail with a fixed direction
 function build_rail(player, start_pos, end_pos, build_last_node)
 	local delta_pos = vector.subtract(end_pos, start_pos)
 	local rail_node_params_seq = advtrain_helpers.direction_step_to_rail_params_sequence(delta_pos)
 	local direction_delta = delta_to_dir(delta_pos)
 	local current_pos = table.copy(start_pos)
-	current_pos.y = current_pos.y - 0.2 -- decrement a little bit, because y values will be rounded for slope placement
 	local tunnelmaker_horizontal_direction = advtrain_helpers.direction_delta_to_advtrains_conn(direction_delta)
 	
 	-- skip first pos if there is already a rail
 	local start_on_rail = advtrain_helpers.is_advtrains_rail_at_pos_or_below(start_pos)
-	if advtrain_helpers.is_advtrains_rail_at_pos_or_below(start_pos) then
```

Comment:
These lines should not be removed unless there is another mechanism in place to skip the first rail if it already exists.
Eventually one could call try_bend_rail_start after placing the next rail...
Of course, the two calls to is_advtrains_rail_at_pos_or_below are redundant and the line above could simply be removed.

Reply:
I wrote the `can_overwrite_track_at` function to check whether it is appropriate to replace the tracks which would, when I finish the PR, make an explicit call to `is_advtrains_rail_at_pos_or_below` on the first track redundant.

My idea with the `can_overwrite_track_at` function is to make it (eventually able to) adjust the track in such a way that it fits the existing setup, such as by changing the track at a particular position to a turnout or a crossing (this behavior is not implemented yet, which is one of the main reasons that this PR is still WIP). This also includes the first track, although I do agree that the possibility of starting at a curve should be taken into consideration.

Reply:
Great, thanks for tackling this. Good to know I don´t have to make up my mind about it.

Reply:
The regression with the first track not being bent should be fixed by 52a636d.